types/src/v0/impls/l1.rs

```diff
@@ -64,180 +75,470 @@ impl L1BlockInfo {
-                sleep(self.retry_delay).await;
-                continue;
-            };
-            if block.number >= number {
-                break block;
+        loop {
+            // Subscribe to events before checking the current state, to ensure we don't miss a relevant
+            // event.
+            let mut events = self.receiver.activate_cloned();
+
+            // Check if the block we are waiting for already exists.
+            {
+                let state = self.state.lock().await;
+                if let Some(finalized) = state.snapshot.finalized {
+                    if finalized.number >= number {
+                        return self.get_finalized_block(state, number).await.1;
+                    }
+                    tracing::info!(
+                        number,
+                        finalized = ?state.snapshot.finalized,
+                        "waiting for l1 finalized block",
+                    );
+                };
             }
-            tracing::info!(current_finalized = %block.number, "waiting for finalized block");
-            sleep(self.retry_delay).await;
-            continue;
-        };
 
-        if finalized.number == number {
-            return finalized;
-        }
+            // Wait for the block.
+            while let Some(event) = events.next().await {
+                let L1Event::NewFinalized { finalized } = event else {
+                    continue;
+                };
+                if finalized.number < number {
+                    tracing::debug!(number, ?finalized, "waiting for L1 finalized block");
+                    continue;
+                }
+                return self
+                    .get_finalized_block(self.state.lock().await, number)
+                    .await
+                    .1;
+            }
 
-        // Load the block again, since it may have changed between first being produced and becoming
-        // finalized.
-        self.wait_for_block(number).await
+            // This should not happen: the event stream ended. All we can do is try again.
+            tracing::warn!(number, "L1 event stream ended unexpectedly; retry",);
+            sleep(self.retry_delay).await;
+        }
     }
 
     /// Get information about the first finalized block with timestamp greater than or equal
     /// `timestamp`.
     pub async fn wait_for_finalized_block_with_timestamp(&self, timestamp: U256) -> L1BlockInfo {
-        let interval = self.provider.get_interval();
-
-        // Sleep until approximately the right time for the desired block to appear.
-        let now = U256::from(OffsetDateTime::now_utc().unix_timestamp() as u64);
-        if timestamp > now {
-            let dur = (timestamp - now).as_u64();
-            tracing::warn!("sleeping for {dur:?}, until {timestamp}");
-            sleep(Duration::from_secs(dur)).await;
-        }
+        // Wait until the finalized block has timestamp >= `timestamp`.
```

Comment:
Wondering if we actually need this function. It's only used on startup if the genesis block is not finalized.

https://github.com/EspressoSystems/espresso-sequencer/blob/b712b84c502d3fe74a530808e40bd2f870eb1d94/sequencer/src/lib.rs#L483

I'm wondering if it would be okay to just panic in that case.

Reply:
No, we can't just panic there. The intended use case is you set the genesis block to something in the future, then you can start up all the nodes ahead of time, and they will just wait until the genesis block is ready and then start running